File: app/src/main/java/org/hackillinois/android/view/schedule/DayFragment.java
package org.hackillinois.android.view.schedule;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.hackillinois.android.R;
import org.hackillinois.android.database.entity.Event;
import org.hackillinois.android.viewmodel.ScheduleViewModel;

import java.util.List;

public class DayFragment extends Fragment {
    private static final String ARG_SECTION_NUM = "section_number";

    private RecyclerView recyclerView;
    private RecyclerView.Adapter adapter;
    private RecyclerView.LayoutManager layoutManager;
    private List<Event> sortedEvents;

    public static DayFragment newInstance(int sectionNumber) {
        DayFragment fragment = new DayFragment();
        Bundle args = new Bundle();

        args.putInt(ARG_SECTION_NUM, sectionNumber);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (adapter != null) adapter.notifyDataSetChanged();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        int sectionNumber = getArguments() == null ? 0 : getArguments().getInt(ARG_SECTION_NUM);

        ScheduleViewModel viewModel = ViewModelProviders.of(this).get(ScheduleViewModel.class);
        viewModel.init();

        LiveData<List<Event>> liveData;

        switch (sectionNumber) {
            case 0: liveData = viewModel.getFridayEventsLiveData();
                break;
            case 1: liveData = viewModel.getSaturdayEventsLiveData();
                break;
            default: liveData = viewModel.getSundayEventsLiveData();
        }

        liveData.observe(this, new Observer<List<Event>>() {
            @Override
            public void onChanged(List<Event> events) {
                sortedEvents = events;
                adapter = new EventsAdapter(sortedEvents);
                recyclerView.setAdapter(adapter);
            }
        });
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_schedule_day, container, false);

        recyclerView = view.findViewById(R.id.activity_schedule_recyclerview);
        layoutManager = new LinearLayoutManager(getContext());
        recyclerView.setLayoutManager(layoutManager);

        if (sortedEvents != null) {
            adapter = new EventsAdapter(sortedEvents);
            recyclerView.setAdapter(adapter);
        }
        return view;
    }
}

File: app/src/main/java/org/hackillinois/android/view/schedule/DayFragment.kt
package org.hackillinois.android.view.schedule

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_schedule_day.view.*
import org.hackillinois.android.R
import org.hackillinois.android.common.FavoritesManager
import org.hackillinois.android.database.entity.Event
import org.hackillinois.android.database.entity.Shift
import org.hackillinois.android.view.MainActivity
import org.hackillinois.android.viewmodel.ScheduleViewModel

class DayFragment : Fragment(), EventClickListener {

    private lateinit var recyclerView: RecyclerView
    private lateinit var mAdapter: EventsAdapter
    private lateinit var mLayoutManager: RecyclerView.LayoutManager

    private var currentEvents: List<Event> = listOf()
    private var currentShifts: List<Shift> = listOf()
    private var showFavorites: Boolean = false
    private var showShifts: Boolean = false
    private var isAttendeeViewing: Boolean = false

    private var listState: Parcelable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val sectionNumber = arguments?.getInt(ARG_SECTION_NUM) ?: 0
        val viewModel = parentFragment?.let { ViewModelProvider(it).get(ScheduleViewModel::class.java) }

        val liveEventData = when (sectionNumber) {
            0 -> viewModel?.fridayEventsLiveData
            1 -> viewModel?.saturdayEventsLiveData
            2 -> viewModel?.sundayEventsLiveData
            else -> viewModel?.fridayEventsLiveData
        }

        isAttendeeViewing = viewModel?.isAttendeeViewing ?: true
        val isLoaded = viewModel?.isLoaded?.value ?: false
        mAdapter = EventsAdapter(listOf(), this, isAttendeeViewing, isLoaded)
        recyclerView = view.activity_schedule_recyclerview.apply {
            mLayoutManager = LinearLayoutManager(context)
            this.layoutManager = mLayoutManager
            this.adapter = mAdapter
        }

        if (currentEvents.isEmpty()) {
            view.postDelayed({
                liveEventData?.observe(
                    this,
                    Observer { events ->
                        events?.let {
                            if (currentEvents.isEmpty() && (isAttendeeViewing || !showShifts)) {
                                currentEvents = it
                                updateEvents(currentEvents)
                                viewModel?.isLoaded?.postValue(true)
                            }
                        }
                    },
                )
            }, 100L)
        } else {
            updateEvents(currentEvents)
            viewModel?.isLoaded?.postValue(true)
        }

        if (isStaff()) {
            val liveShiftData = when (sectionNumber) {
                0 -> viewModel?.fridayShiftsLiveData
                1 -> viewModel?.saturdayShiftsLiveData
                2 -> viewModel?.sundayShiftsLiveData
                else -> viewModel?.fridayShiftsLiveData
            }
            viewModel?.showShifts?.observe(
                this,
                Observer {
                    showShifts = it
                    if (showShifts) {
                        // Log.d("Observe showShifts", "Switching to SHIFTS")
                        mAdapter.animateItems = false
                        mAdapter.updateEvents(insertTimeItems(currentShifts))
                    } else {
                        // Log.d("Observe showShifts", "Switching to SCHEDULE")
                        updateEvents(currentEvents)
                    }
                },
            )
            view.postDelayed({
                liveShiftData?.observe(
                    this,
                    Observer { shifts ->
                        shifts?.let {
                            currentShifts = it
                            if (showShifts) {
                                mAdapter.updateEvents(insertTimeItems(currentShifts))
                            }
                        }
                    },
                )
            }, 100)
        }

        if (isAttendeeViewing) {
            viewModel?.showFavorites?.observe(
                this,
                Observer {
                    showFavorites = it
                    updateEvents(currentEvents)
                },
            )
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        val view = inflater.inflate(R.layout.fragment_schedule_day, container, false)
        return view
    }

    override fun onSaveInstanceState(state: Bundle) {
        super.onSaveInstanceState(state)
        // Save list state
        state.putParcelable("recyclerview", mLayoutManager.onSaveInstanceState())
    }

    override fun openEventInfoActivity(event: Event) {
        val eventInfoFragment = EventInfoFragment.newInstance(event.eventId, isAttendeeViewing)
        (activity as MainActivity?)?.switchFragment(eventInfoFragment, true)
    }

    private fun updateEvents(list: List<Event>) {
        var listTemp = list
        context?.let {
            if (showFavorites) {
                listTemp = listTemp.filter { event -> FavoritesManager.isFavoritedEvent(it, event.eventId) }
            }
        }
        mAdapter.updateEvents(insertTimeItems(listTemp))
    }

    private fun insertTimeItems(eventList: List<ScheduleListItem>): List<ScheduleListItem> {
        var currentTime = -1L
        val newList = mutableListOf<ScheduleListItem>()

        eventList.forEach {
            if (it.getStartTimeMs() != currentTime) {
                currentTime = it.getStartTimeMs()
                val timeListItem = TimeListItem(it.getStartTimeOfDay())
                newList.add(timeListItem)
            }
            newList.add(it)
        }

        return newList
    }

    private fun isStaff(): Boolean {
        val context = requireActivity().applicationContext
        val prefString = context.getString(R.string.authorization_pref_file_key)
        return context.getSharedPreferences(prefString, Context.MODE_PRIVATE).getString("provider", "") ?: "" == "google"
    }

    companion object {
        private val ARG_SECTION_NUM = "section_number"

        fun newInstance(sectionNumber: Int): DayFragment {
            val fragment = DayFragment()
            val args = Bundle()

            args.putInt(ARG_SECTION_NUM, sectionNumber)
            fragment.arguments = args

            return fragment
        }
    }
}
